frontend/packages/topology/src/data-transforms/DataModelProvider.tsx

```diff
@@ -19,30 +17,64 @@ interface DataModelProviderProps {
   children?: React.ReactNode;
 }
 
-export const getNamespacedDynamicModelFactories = (
+export const omitResources = (
   factories: LoadedExtension<DynamicTopologyDataModelFactory>[],
-) =>
-  factories.map(({ properties, ...ext }) => {
-    return {
-      ...ext,
-      properties: {
-        ...properties,
-        resources: (namespace: string) =>
-          Object.assign(
-            {},
-            ...Object.entries(properties.resources).map(([k, v]) => {
-              const kind = v?.model?.version
-                ? referenceForExtensionModel(v.model)
-                : v?.model
-                ? referenceForModel(modelForGroupKind(v.model?.group, v.model?.kind))
-                : v?.opts?.kind;
-
-              return { [k]: { namespace, kind, ...v?.opts } };
-            }),
-          ),
-      },
-    };
+): LoadedExtension<Omit<DynamicTopologyDataModelFactory, 'resources'>>[] => {
+  const otherProperties = factories.map(({ properties, ...ext }) => {
+    const { resources, ...other } = properties;
+    return { properties: other, ...ext };
   });
+  return otherProperties;
+};
+
+const WatchResourcesHookResolver: React.FC<{
+  namespace: string;
+  useResources: WatchResourcesHook;
+  onResolve: (resources: WatchK8sResources<any>) => void;
+}> = ({ namespace, useResources, onResolve }) => {
+  const resources = useResources(namespace);
+  React.useEffect(() => {
+    onResolve(resources);
+  }, [onResolve, resources]);
+  return null;
+};
+
+export const DynamicWatchResourcesLoader: React.FC<{
+  namespace: string;
+  children: (props: WatchK8sResources<any>) => React.ReactNode;
+}> = ({ namespace, children }) => {
+  const [watchResources, setWatchResources] = React.useState<WatchK8sResources<any>>({});
+  const [modelFactories, loaded] = useResolvedExtensions<DynamicTopologyDataModelFactory>(
+    isDynamicTopologyDataModelFactory,
+  );
+  const modelResources = React.useMemo(
+    () =>
+      modelFactories
+        ?.sort((r1, r2) => r1.properties.priority - r2.properties.priority)
+        .map((m) => ({ id: m.properties.id, resourceHook: m.properties.resources }))
+        .filter(({ resourceHook }) => !!resourceHook),
+    [modelFactories],
+  );
+
+  return (
+    <>
+      <>
+        {loaded &&
+          modelResources.map(({ id, resourceHook }) => (
+            <WatchResourcesHookResolver
+              key={id}
+              namespace={namespace}
+              useResources={resourceHook}
+              onResolve={(resources) => {
```

Comment:
This could result in maximum depth reached error when hooks are asynchronous and have side effects.